### DISCUSSION TITLE
How to add marks dynamically

### DISCUSSION THREAD

**Discussion by M1s4k1** [pytest-dev/pytest#11743] [Q&A] [answered]
In pytest, I attempted to dynamically add @pytest.mark.xxx to test cases using @env_mark. However, it doesn't take effect on test cases decorated with pytest.mark.parametrize. 
I suspect it's related to the lifecycle of test cases dynamically generated by @parametrize. How should I modify my code?

Below is a brief example of the test case code:

```python
class TestOrder(HttpBaseTestCase):
    @env_mark(["dev", "test"])
    @pytest.mark.parametrize('order_id', ['123', '456'])
    def test_do_get_order_id(self, order_id):
        pass
```

Here is the simplified @env_mark code:

```python
def env_mark(env):

    def _env_mark(func):
        def wrapper(*args, **kwargs):
            return func(*args, **kwargs)

        env_wrapper = [_mark_map[i] for i in env]
        for _env_wrapper in env_wrapper:
            wrapper = _env_wrapper(wrapper)

        return functools.wraps(func)(wrapper)

    return _env_mark

_mark_map = {
    "dev": pytest.mark.dev,
    "test": pytest.mark.,
    "prod": pytest.mark.prod
and so on...
}
```

**Comment by RonnyPfannschmidt** [comment, marked as answer]
Please stop creating all those wrappers

Instead just directly apply the Marks

**Comment by sneakers-the-rat** [comment]
So I have a non-wrapper related need for this:

I am doing combinatoric testing across a number of parameters. Only some of those combinations are valid. I would like to provide the parameterization as a fixture so that i can repeat the parameterization across a number of tests. I want to mark the tests according to one of the parameters. The closest I can come is something like this:

```python
from itertools import product

cases = itertools.product([1,2,3], ['a', 'b', 'c'])

@pytest.fixture(
    params = [
        pytest.param('x', marks=pytest.mark.x), 
        pytest.param('y', marks=pytest.mark.y), 
        pytest.param('z', marks=pytest.mark.y)
    ]
)
def other_cases(request):
    return request.param

@pytest.fixture(params=cases)
def combinatoric_fixture(other_cases, request):
    # assume this can be done
    case = merge(other_cases, request.param)
    if case.skip():
        pytest.skip()
    return case
```

That allows me to mark the tests correctly, but it produces a ton of skips, and since this is a product over a large number of cases that ends up burning a decent amount of time because there is a ~small amount of work done when iterating to merge the results of the product.

I have tried using `request.applymarker` and `request._pyfuncitem.add_marker` to no effect. I have also tried using several different fixtures for each of the things i'm iterating over, but it doesn't seem to be possible to combine multiple parameterized fixtures in another fixture.

It would actually be really great to be able to apply multiple marks for each of the items in the product, but i will settle for just one. Is there some way of doing this that i'm missing?

**Reply by sneakers-the-rat** [reply]
I have found a solution to this, If i use a generator expression like this:

```python
@pytest.fixture(
    params=(
        pytest.param(p, id=p.id, marks=getattr(pytest.mark, p.name))
        for p in ALL_CASES_PASSING
    )
)
def all_cases_passing(request):
    return request.param
```

Then it works. it makes for a little uglier declaration and a little more awkwardness in how i write the fixtures but that's good enough for me.
